camel/memories/blocks/graphdb_block.py

```diff
@@ -0,0 +1,176 @@
+
+    def retrieve(
+        self, query: str, params: Optional[Dict[str, Any]] = None
+    ) -> List[ContextRecord]:
+        r"""Retrieves records from the graph database based on a query.
+
+        Args:
+            query (str): The query string to execute.
+            params (Optional[Dict[str, Any]]): Parameters for the query.
+
+        Returns:
+            List[ContextRecord]: A list of context records retrieved from the
+                graph database.
+        """
+        try:
+            results = self.storage.query(query, params)
+            scores = self._calculate_scores(query, results)
+            return [
+                ContextRecord(
+                    memory_record=MemoryRecord.from_dict(result), score=score
+                )
+                for result, score in zip(results, scores)
+            ]
+        except Exception as e:
+            logger.error(f"Error retrieving records: {e}")
+            return []
+
+    def write_records(self, records: List[MemoryRecord]) -> None:
+        """Writes records to the graph database."""
+        for record in records:
+            content = record.message.content
+
+            if isinstance(content, str):
+                try:
+                    data = json.loads(content)
+                    subj = data.get('subject')
+                    obj = data.get('object')
+                    rel = data.get('relation')
+                except json.JSONDecodeError:
+                    logger.error(f"Error parsing content as JSON: {content}")
+                    continue
+            else:
+                subj = content.get('subject')
+                obj = content.get('object')
+                rel = content.get('relation')
+
+            if subj and obj and rel:
+                self.write_triplet(subj, obj, rel)
+
+    def write_triplet(self, subj: str, obj: str, rel: str) -> None:
+        r"""Writes a triplet to the graph database.
+
+        Args:
+            subj (str): The subject of the triplet.
+            obj (str): The object of the triplet.
+            rel (str): The relationship between subject and object.
+        """
+        self.storage.add_triplet(subj, obj, rel)
+
+    def delete_triplet(self, subj: str, obj: str, rel: str) -> None:
+        r"""Deletes a triplet from the graph database.
+
+        Args:
+            subj (str): The subject of the triplet.
+            obj (str): The object of the triplet.
+            rel (str): The relationship between subject and object.
+        """
+        self.storage.delete_triplet(subj, obj, rel)
+
+    def _calculate_scores(self, query: str, results: List[Any]) -> List[float]:
+        query_graph = self._build_graph(query)
+        scores = []
+        for result in results:
+            result_graph = self._build_graph(result)
+            score = self._graph_similarity(query_graph, result_graph)
+            scores.append(score)
+        return scores
```

Comment:
for the score I think using an embedding-based vector would be more effective for calculating the score, as it aligns better with matching the user's query. A semantic-based score would be more suitable in this case. WDYT?